### ISSUE TITLE
Can't parse complex 'between' statement

### ISSUE THREAD

**Issue by pwildenhain** [sqlfluff/sqlfluff#478]
test.sql
```sql
select
    my_table.*
from
    my_table
    inner join my_other_table
        on my_table.key = my_other_table.key
where
    my_table.field_1 between my_other_table.field_2 - 10 and my_other_table.field_3
```

lint result
```sh
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   8 | P:  21 |  PRS | Found unparsable section: ' between
                       | my_other_table.field_2 - 10 and...'
WARNING: Parsing errors found and dialect is set to 'ansi'. Have you configured your dialect?
```

This also happens if the complex `between` statement is in a `join` statement. The issue goes away if simplify  `my_other_table.field_2 - 10 ` to just `my_other_table.field_2`

**Comment by alanmcruickshank** [comment]
The `between` expression is currently defined in `Expression_A_Grammar` and *should* support more complex expressions within it. Evidently something is going wrong with that. If you run `sqlfluff parse` on the query you posted what does it return?

From the error message it looks like it's consuming `where my_table.field_1` as the where clause and cutting off the `between ...`. If we can find out what it's parsing the first bit as, then it might lead us to why the latter isn't being included for some reason.

**Comment by pwildenhain** [comment]
Here's the parse output of the query:
```
FileSegment:                                                [0](1, 1, 1)
    StatementSegment:                                       [0](1, 1, 1)
        SelectStatementSegment:                             [0](1, 1, 1)
            SelectClauseSegment:                            [0](1, 1, 1)
                SELECT_KeywordSegment:                      [0](1, 1, 1)        'select'
                [META] Indent:                              [6](1, 1, 7)
                newline_RawSegment:                         [6](1, 1, 7)        '\n'
                whitespace_RawSegment:                      [7](1, 2, 1)        '    '
                SelectTargetElementSegment:                 [11](1, 2, 5)
                    WildcardExpressionSegment:              [11](1, 2, 5)
                        WildcardIdentifierSegment:          [11](1, 2, 5)
                            naked_identifier_ReSegment:     [11](1, 2, 5)       'my_table'
                            dot_KeywordSegment:             [19](1, 2, 13)      '.'
                            star_KeywordSegment:            [20](1, 2, 14)      '*'
                [META] Dedent:                              [21](1, 2, 15)
            newline_RawSegment:                             [21](1, 2, 15)      '\n'
            FromClauseSegment:                              [22](1, 3, 1)
                FROM_KeywordSegment:                        [22](1, 3, 1)       'from'
                [META] Indent:                              [26](1, 3, 5)
                whitespace_RawSegment:                      [26](1, 3, 5)       ' '
                TableExpressionSegment:                     [27](1, 3, 6)
                    ObjectReferenceSegment:                 [27](1, 3, 6)
                        naked_identifier_ReSegment:         [27](1, 3, 6)       'my_table'
                newline_RawSegment:                         [35](1, 3, 14)      '\n'
                [META] Dedent:                              [36](1, 4, 1)
                JoinClauseSegment:                          [36](1, 4, 1)
                    INNER_KeywordSegment:                   [36](1, 4, 1)       'inner'
                    whitespace_RawSegment:                  [41](1, 4, 6)       ' '
                    JOIN_KeywordSegment:                    [42](1, 4, 7)       'join'
                    [META] Indent:                          [46](1, 4, 11)
                    whitespace_RawSegment:                  [46](1, 4, 11)      ' '
                    TableExpressionSegment:                 [47](1, 4, 12)
                        ObjectReferenceSegment:             [47](1, 4, 12)
                            naked_identifier_ReSegment:     [47](1, 4, 12)      'my_other_table'
                    newline_RawSegment:                     [61](1, 4, 26)      '\n'
                    whitespace_RawSegment:                  [62](1, 5, 1)       '    '
                    ON_KeywordSegment:                      [66](1, 5, 5)       'on'
                    [META] Indent:                          [68](1, 5, 7)
                    whitespace_RawSegment:                  [68](1, 5, 7)       ' '
                    ExpressionSegment:                      [69](1, 5, 8)
                        ObjectReferenceSegment:             [69](1, 5, 8)
                            naked_identifier_ReSegment:     [69](1, 5, 8)       'my_table'
                            dot_KeywordSegment:             [77](1, 5, 16)      '.'
                            naked_identifier_ReSegment:     [78](1, 5, 17)      'key'
                        whitespace_RawSegment:              [81](1, 5, 20)      ' '
                        equals_KeywordSegment:              [82](1, 5, 21)      '='
                        whitespace_RawSegment:              [83](1, 5, 22)      ' '
                        ObjectReferenceSegment:             [84](1, 5, 23)
                            naked_identifier_ReSegment:     [84](1, 5, 23)      'my_other_table'
                            dot_KeywordSegment:             [98](1, 5, 37)      '.'
                            naked_identifier_ReSegment:     [99](1, 5, 38)      'key'
                    [META] Dedent:                          [102](1, 5, 41)
                    [META] Dedent:                          [102](1, 5, 41)
            newline_RawSegment:                             [102](1, 5, 41)     '\n'
            WhereClauseSegment:                             [103](1, 6, 1)
                WHERE_KeywordSegment:                       [103](1, 6, 1)      'where'
                [META] Indent:                              [108](1, 6, 6)
                newline_RawSegment:                         [108](1, 6, 6)      '\n'
                whitespace_RawSegment:                      [109](1, 7, 1)      '    '
                ExpressionSegment:                          [113](1, 7, 5)
                    ObjectReferenceSegment:                 [113](1, 7, 5)
                        naked_identifier_ReSegment:         [113](1, 7, 5)      'my_table'
                        dot_KeywordSegment:                 [121](1, 7, 13)     '.'
                        naked_identifier_ReSegment:         [122](1, 7, 14)     'field_1'
                [META] Dedent:                              [129](1, 7, 21)
                UnparsableSegment:                          [129](1, 7, 21)     !! Expected: 'Nothing...'
                    whitespace_RawSegment:                  [129](1, 7, 21)     ' '
                    code_RawSegment:                        [130](1, 7, 22)     'between'
                    whitespace_RawSegment:                  [137](1, 7, 29)     ' '
                    code_RawSegment:                        [138](1, 7, 30)     'my_other_table'
                    dot_RawSegment:                         [152](1, 7, 44)     '.'
                    code_RawSegment:                        [153](1, 7, 45)     'field_2'
                    whitespace_RawSegment:                  [160](1, 7, 52)     ' '
                    minus_RawSegment:                       [161](1, 7, 53)     '-'
                    whitespace_RawSegment:                  [162](1, 7, 54)     ' '
                    numeric_literal_RawSegment:             [163](1, 7, 55)     '10'
                    whitespace_RawSegment:                  [165](1, 7, 57)     ' '
                    code_RawSegment:                        [166](1, 7, 58)     'and'
                    whitespace_RawSegment:                  [169](1, 7, 61)     ' '
                    code_RawSegment:                        [170](1, 7, 62)     'my_other_table'
                    dot_RawSegment:                         [184](1, 7, 76)     '.'
                    code_RawSegment:                        [185](1, 7, 77)     'field_3'
    newline_RawSegment:
```

And here is the parse of the query **without** the complex between. This one actually catches the `between` while the other fails at that section
```
FileSegment:                                                [0](1, 1, 1)
    StatementSegment:                                       [0](1, 1, 1)
        SelectStatementSegment:                             [0](1, 1, 1)
            SelectClauseSegment:                            [0](1, 1, 1)
                SELECT_KeywordSegment:                      [0](1, 1, 1)        'select'
                [META] Indent:                              [6](1, 1, 7)
                newline_RawSegment:                         [6](1, 1, 7)        '\n'
                whitespace_RawSegment:                      [7](1, 2, 1)        '    '
                SelectTargetElementSegment:                 [11](1, 2, 5)
                    WildcardExpressionSegment:              [11](1, 2, 5)
                        WildcardIdentifierSegment:          [11](1, 2, 5)
                            naked_identifier_ReSegment:     [11](1, 2, 5)       'my_table'
                            dot_KeywordSegment:             [19](1, 2, 13)      '.'
                            star_KeywordSegment:            [20](1, 2, 14)      '*'
                [META] Dedent:                              [21](1, 2, 15)
            newline_RawSegment:                             [21](1, 2, 15)      '\n'
            FromClauseSegment:                              [22](1, 3, 1)
                FROM_KeywordSegment:                        [22](1, 3, 1)       'from'
                [META] Indent:                              [26](1, 3, 5)
                whitespace_RawSegment:                      [26](1, 3, 5)       ' '
                TableExpressionSegment:                     [27](1, 3, 6)
                    ObjectReferenceSegment:                 [27](1, 3, 6)
                        naked_identifier_ReSegment:         [27](1, 3, 6)       'my_table'
                newline_RawSegment:                         [35](1, 3, 14)      '\n'
                [META] Dedent:                              [36](1, 4, 1)
                JoinClauseSegment:                          [36](1, 4, 1)
                    INNER_KeywordSegment:                   [36](1, 4, 1)       'inner'
                    whitespace_RawSegment:                  [41](1, 4, 6)       ' '
                    JOIN_KeywordSegment:                    [42](1, 4, 7)       'join'
                    [META] Indent:                          [46](1, 4, 11)
                    whitespace_RawSegment:                  [46](1, 4, 11)      ' '
                    TableExpressionSegment:                 [47](1, 4, 12)
                        ObjectReferenceSegment:             [47](1, 4, 12)
                            naked_identifier_ReSegment:     [47](1, 4, 12)      'my_other_table'
                    newline_RawSegment:                     [61](1, 4, 26)      '\n'
                    whitespace_RawSegment:                  [62](1, 5, 1)       '    '
                    ON_KeywordSegment:                      [66](1, 5, 5)       'on'
                    [META] Indent:                          [68](1, 5, 7)
                    whitespace_RawSegment:                  [68](1, 5, 7)       ' '
                    ExpressionSegment:                      [69](1, 5, 8)
                        ObjectReferenceSegment:             [69](1, 5, 8)
                            naked_identifier_ReSegment:     [69](1, 5, 8)       'my_table'
                            dot_KeywordSegment:             [77](1, 5, 16)      '.'
                            naked_identifier_ReSegment:     [78](1, 5, 17)      'key'
                        whitespace_RawSegment:              [81](1, 5, 20)      ' '
                        equals_KeywordSegment:              [82](1, 5, 21)      '='
                        whitespace_RawSegment:              [83](1, 5, 22)      ' '
                        ObjectReferenceSegment:             [84](1, 5, 23)
                            naked_identifier_ReSegment:     [84](1, 5, 23)      'my_other_table'
                            dot_KeywordSegment:             [98](1, 5, 37)      '.'
                            naked_identifier_ReSegment:     [99](1, 5, 38)      'key'
                    [META] Dedent:                          [102](1, 5, 41)
                    [META] Dedent:                          [102](1, 5, 41)
            newline_RawSegment:                             [102](1, 5, 41)     '\n'
            WhereClauseSegment:                             [103](1, 6, 1)
                WHERE_KeywordSegment:                       [103](1, 6, 1)      'where'
                [META] Indent:                              [108](1, 6, 6)
                newline_RawSegment:                         [108](1, 6, 6)      '\n'
                whitespace_RawSegment:                      [109](1, 7, 1)      '    '
                ExpressionSegment:                          [113](1, 7, 5)
                    ObjectReferenceSegment:                 [113](1, 7, 5)
                        naked_identifier_ReSegment:         [113](1, 7, 5)      'my_table'
                        dot_KeywordSegment:                 [121](1, 7, 13)     '.'
                        naked_identifier_ReSegment:         [122](1, 7, 14)     'field_1'
                    whitespace_RawSegment:                  [129](1, 7, 21)     ' '
                    BETWEEN_KeywordSegment:                 [130](1, 7, 22)     'between'
                    whitespace_RawSegment:                  [137](1, 7, 29)     ' '
                    ObjectReferenceSegment:                 [138](1, 7, 30)
                        naked_identifier_ReSegment:         [138](1, 7, 30)     'my_other_table'
                        dot_KeywordSegment:                 [152](1, 7, 44)     '.'
                        naked_identifier_ReSegment:         [153](1, 7, 45)     'field_2'
                    whitespace_RawSegment:                  [160](1, 7, 52)     ' '
                    AND_KeywordSegment:                     [161](1, 7, 53)     'and'
                    whitespace_RawSegment:                  [164](1, 7, 56)     ' '
                    ObjectReferenceSegment:                 [165](1, 7, 57)
                        naked_identifier_ReSegment:         [165](1, 7, 57)     'my_other_table'
                        dot_KeywordSegment:                 [179](1, 7, 71)     '.'
                        naked_identifier_ReSegment:         [180](1, 7, 72)     'field_3'
                [META] Dedent:                              [187](1, 7, 79)
    newline_RawSegment: 
```

**Comment by alanmcruickshank** [comment]
@pwildenhain - in debugging this further I actually fixed this. I hope you've not invested a ton of time into trying to fix it. PR coming shortly. Take a look at the PR I'm about to put up.

**Comment by pwildenhain** [comment]
Oh fabulous 👯‍♂️ nope I hadn't started working on it yet
